### ISSUE TITLE
텍스트 분류 - [책]김기현의 자연어 처리 딥러닝 캠프 파이토치편

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/NLU-NLP#9]
## 텍스트 분류 Task
: 텍스트를 입력으로 받아, **불연속적 값(심볼)**으로 출력하는 문제 대부분에 해당. (예: 감성분석, 스팸메일 탐지, 사용자 의도분류, 주제 분류, 카테고리 분류 ..)
-(목적) 문장을 latent space(잠재공간)에 projection하여 decision boundary를 찾는 것.
(= 확률적 관점으로는, 문장이 주어졌을 때, 문장이 속할 클래스의 확률 분포 함수를 approximate 하는 것.)
-(traditional 기법) 나이브 베이즈 분류, SVM 등 ..
<img width="858" alt="스크린샷 2021-01-20 오후 2 59 28" src="https://user-images.githubusercontent.com/70500841/105133493-3b840e00-5b30-11eb-8176-6a00903d2e65.png">

## 나이브 베이즈 (traditional 기법)
: 간단한 가정으로 데이터의 희소성 문제를 해결하는 쉬우면서도 강력한 방법.

-(목적) **사후 확률을 최대화**하는 클래스 예측하는 **MAP**(Maximum a posteriori) 기반.
- 사후확률 = n개 단어 w1, w2, ..., wn이 주어졌을 때, 문장이 c 클래스에 속할 확률값

-(기존 문제) MAP에서는 **사후확률(posterior) ∝ 우도(likelihood) X 사전확률(prior)** 을 통해 클래스 y를 예측.
하지만, Dataset이 복잡한 feature 특징으로 이루어진 데이터라면(**텍스트에서는 문장이 복잡하거나 길수록 복잡한 feature**), 확률은 코퍼스의 출현빈도로 추정가능한데, 훈련 데이터에서 매우 희소하기에 => 코퍼스에 없는 특징 조합이라는 이유로 확률값을 0으로 추정해야 함 => risky

-(가정) 각 특징이 독립적이라고 가정. => **각 특징의 결합확률은 각 독립된 확률의 곱**으로 근사.
- P(y=c | x = w1, w2, ..., wn) ∝ P(x = w1, w2, ..., wn | y=c) X P(y=c)
    ≈ P(w1|c) P(w2|c) ... P(wn|c) X P(c)
    = Π (i=1...n) P(wi|c) X P(c) 

=> 따라서,  **각 특징들의 확률의 곱에 사전확률을 곱한 값을 최대화하는 클래스 예측** 문제로 해결 가능 !
<img width="933" alt="스크린샷 2021-01-20 오후 3 30 19" src="https://user-images.githubusercontent.com/70500841/105136258-b2bba100-5b34-11eb-8dbc-65e4cd70e549.png">

-1) **사전확률**(prior) : **Dataset(코퍼스)에서 출현한 빈도**를 통해 추정 -> _해당 class 개수 / 전체 class 개수 ?_
-2) 각 특징별 **우도**(likelihood) : **Dataset(코퍼스)에서 출현한 빈도**를 통해 추정 -> _해당 단어의 출현빈도 / 모든 단어의 출현빈도의 합 ?_
<img width="278" alt="스크린샷 2021-01-20 오후 3 39 02" src="https://user-images.githubusercontent.com/70500841/105136829-a71caa00-5b35-11eb-82a6-6f4d12197212.png">

- 지난 스터디 발표 이야기 생각.. 불량/이상 탐지 문제에서 실제 불량은 1~3%로 매우 적기에 class imbalance를 보완하기 위한 가정 -> 불량 : 정상 비율 1:9 맞추기 등,,

#### add-one Smoothing
- 각 출현횟수에 1을 더해주는 보정 기법
- 우도(likelihood) = _(해당 단어의 출현횟수 + 1) / (모든 단어의 출현횟수 + |Vocab Size|)_ 이 됨

----------------------------------
###### MLE (Maximum Likelihood Estimation, 최대 가능도 추정)
- 실제 확률분포로부터 샘플링한, 수집된 데이터를 가장 잘 설명하는(또는 결과값을 잘 예측하는) 확률 분포 모델을 추정하는 '일반화'과정이 머신러닝
- Likelihood : 확률분포의 파라미터(θ 등)가 데이터를 얼마나 잘 설명하는지를(=확률밀도 값이 크거나, 확률값 자체가 크거나) 나타내는 정도
- Likelihood Function : 입력으로 주어진 확률분포의 "파라미터(θ 등)"가 데이터를 얼마나 잘 설명하는지를(=확률밀도 값이 크거나, 확률값 자체가 크거나) 나타내는 정도인, "Likelihood"를 출력하는 함수

- Likelihood는 확률값(확률밀도 값)의 **곱** -> Log를 취해 **합**으로 표현 가능(underflow 방지 효과, 연산 빠름) -> Gradient Ascent를 통해 이를 Maximize 해야하지만, (-)를 취해 NLL(Negative Log-Likelihood)로 Gradient Descent 를 통해 Minimize함으로써 -> 최종적으로 Likelihood 값을 MAX 최대로 만드는 파라미터(θ 등)를 찾을 수 있다 !
<img width="243" alt="스크린샷 2021-02-03 오후 3 03 50" src="https://user-images.githubusercontent.com/70500841/106705521-7490a800-6631-11eb-9edd-aabc2b075d2b.png">
<img width="235" alt="스크린샷 2021-02-03 오후 3 06 47" src="https://user-images.githubusercontent.com/70500841/106705523-765a6b80-6631-11eb-8d8a-9024d35fa680.png">

###### MAP (Maximum A Posterior, 사후확률 최대화)
- 데이터가 주어졌을 때 사후확률(Posterior)를 최대화하는 가설(hypothesis)를 찾아내는 방법
<img width="416" alt="스크린샷 2021-02-03 오후 3 11 52" src="https://user-images.githubusercontent.com/70500841/106705944-30ea6e00-6632-11eb-806f-ee983ef51683.png">
- P(D)는 h와 관련 없으므로 생략하면 -> 즉, likelihood X prior 를 최대화 하는 과정과 같아짐
- 가설 h는 파라미터(θ 등)에 해당 !

* 베이즈 정리
<img width="436" alt="스크린샷 2021-02-03 오후 3 08 11" src="https://user-images.githubusercontent.com/70500841/106705633-aa359100-6631-11eb-8132-1b9a85782f26.png">

![image](https://user-images.githubusercontent.com/70500841/105134744-3a53e080-5b32-11eb-9370-cb2eaf4d3350.png)
- _Class Imbalance와 관련 있는 것 같다? (남자 여자 신발사이즈 예제)_ => 실제로 Bayesian 관점은 미래의 불확실성까지 고려한다는 장점이 있지만 **Prior에 대한 가정**(θ 또한 랜덤변수로, 어떠한 분포를 가지기 때문에)이 필요함 => 이는 Frequentist 관점과 달리 overfitting 문제도 해결할 수 있기 때문에 계속 연구중
